File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>
#include <cmath>
using namespace std;

set<int> combineArr;
//1, 7
//0,1,1 => 01,11,101
void makeCombine(string combines, string others){
    // 새로운 경우의수 넣기
    if(combines != ""){
        combineArr.insert(stoi(combines));
    }
    //기존 경우의수 + 새로운 값
    for (int i = 0; i < others.size(); i++)
        makeCombine(combines + others[i], others.substr(0, i) + others.substr(i + 1));
}
bool isPrime(int number){
    if (number == 0 || number == 1)
        return false;
    
    for (int i = 2; i <= sqrt(number); i++)
        if (number % i == 0)
            return false;

    return true;
}

int solution(string numbers) {
    int answer = 0;
    // 1. 모든 경우의 수를 만든다
    makeCombine("", numbers);
    // 2. 소수인지 판별하고 개수를 +1 한다.
    for(int i : combineArr){
        if(isPrime(i)){
            answer++;
        }
    }
    return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
import java.io.*;
class Solution {
    static Set<Integer> combSet;
    static boolean[] visited = new boolean[7];
    
    public int solution(String numbers) {
        int answer = 0;
        combSet = new HashSet<Integer>();
//         1. 모든 조합을 만든다.
//         2. 조합이 소수인지 카운트 한다.
        
        dfs(numbers, 0, "");
        
        for(Integer i : combSet){
            if(isPrime(i)){
                answer++;
            }
        }   
        
        
        return answer;
    }
    static void dfs(String numbers, int depth, String s){
        if(depth > numbers.length()){
            return;
        }
        
        for(int i = 0; i < numbers.length(); i++){
            if(visited[i] == false){
                visited[i] = true;
                combSet.add(Integer.parseInt(s+numbers.charAt(i)));
                dfs(numbers, depth +1, s + numbers.charAt(i));
                visited[i] = false;
            }
        }
        
    }
    
    static boolean isPrime(int n){
        if(n == 0 || n == 1){
            return false;
        }
        
        for(int i = 2; i <= (int)Math.sqrt(n); i++){
            if(n%i == 0){
                return false;
            }
        }
        return true;
    }
}